pennylane/ops/qubit/observables.py

```diff
@@ -349,54 +347,126 @@ class Projector(Observable):
+    we need to use its state vector form ``plusplus_state=np.array([1, 1, 1, 1])/2``.
+
+    .. code-block::
+
+        >>> dev = qml.device("default.qubit", wires=2)
+        >>> @qml.qnode(dev)
+        ... def circuit(state, wires):
+        ...     return qml.expval(qml.Projector(state, wires=[0, 1]))
+        >>> zero_state = [0, 0]
+        >>> circuit(zero_state, wires)
+        1.
+        >>> plusplus_state = np.array([1, 1, 1, 1]) / 2
+        >>> circuit(plusplus_state, wires)
+        0.25
+
     """
     num_wires = AnyWires
     num_params = 1
     """int: Number of trainable parameters that the operator depends on."""
+    _basis_state_type = None  # type if Projector inherits from _BasisStateProjector
+    _state_vector_type = None  # type if Projector inherits from _StateVectorProjector
+
+    def __new__(cls, state, wires, **_):
+        """Changes parents based on the state representation.
+
+        Though all the types will be named "Projector", their *identity* and location in memory
+        will be different based on whether the input state is a basis state or a state vector.
+        We cache the different types in private class variables so that:
+
+        >>> Projector(state, wires).__class__ is Projector(state, wires).__class__
+        True
+        >>> type(Projector(state, wries)) == type(Projector(state, wires))
+        True
+        >>> isinstance(Projector(state, wires), type(Projector(state, wires)))
+        True
+        >>> Projector(basis_state, wires).__class__ is Projector._basis_state_type
+        True
+        >>> Projector(state_vector, wires).__class__ is Projector._state_vector_type
+        True
 
-    def __init__(self, basis_state, wires, do_queue=None, id=None):
+        """
         wires = Wires(wires)
-        shape = qml.math.shape(basis_state)
-
+        shape = qml.math.shape(state)
         if len(shape) != 1:
-            raise ValueError(f"Basis state must be one-dimensional; got shape {shape}.")
+            raise ValueError(f"Input state must be one-dimensional; got shape {shape}.")
+
+        if len(state) == len(wires):
+            if cls._basis_state_type is None:
+                base_cls = (_BasisStateProjector, Projector)
+                cls._basis_state_type = type("Projector", base_cls, dict(cls.__dict__))
+            return object.__new__(cls._basis_state_type)
+
+        if len(state) == 2 ** len(wires):
+            if cls._state_vector_type is None:
+                base_cls = (_StateVectorProjector, Projector)
+                cls._state_vector_type = type("Projector", base_cls, dict(cls.__dict__))
+            return object.__new__(cls._state_vector_type)
+
+        raise ValueError(
+            "Input state should have the same length as the wires in the case"
+            " of a basis state, or 2**len(wires) in case of a state vector."
+            f" The lengths for the input state and wires are {len(state)} and "
+            f" {len(wires)}, respectively."
```

Comment:
apologies for nit-picking here but we prefer spaces at the end of lines for multi-line strings
```suggestion
            "Input state should have the same length as the wires in the case "
            "of a basis state, or 2**len(wires) in case of a state vector. "
            f"The lengths for the input state and wires are {len(state)} and "
            f"{len(wires)}, respectively."
```

Reply:
No problem! Now that I look at it, it does indeed look odd to have the spaces at the beginning